sdk/python/lib/pulumi/runtime/stack.py

```diff
@@ -92,11 +94,92 @@ def __init__(self, func: Callable) -> None:
         try:
             func()
         finally:
-            self.register_outputs(self.outputs)
+            self.register_outputs(massage(self.outputs, []))
             # Intentionally leave this resource installed in case subsequent async work uses it.
 
     def output(self, name: str, value: Any):
         """
         Export a stack output with a given name and value.
         """
         self.outputs[name] = value
+
+# Note: we use a List here instead of a set as many objects are unhashable.  This is inefficient,
+# but python seems to offer no alternative.
+def massage(attr: Any, seen: List[Any]):
+    """
+    massage takes an arbitrary python value and attempts to *deeply* convert it into
+    plain-old-python-value that can registered as an output.  In general, this means leaving alone
+    things like strings, ints, bools. However, it does mean trying to make other values into either
+    lists or dictionaries as appropriate.  In general, iterable things are turned into lists, and
+    dictionary-like things are turned into dictionaries.
+    """
+
+    # Basic primitive types (numbers, booleans, strings, etc.) don't need any special handling.
+
+    if is_primitive(attr):
+        return attr
+
+    # from this point on, we have complex objects.  If we see them again, we don't want to emit them
+    # again fully or else we'd loop infinitely.
+    if reference_contains(attr, seen):
+        # Note: for Resources we hit again, emit their urn so cycles can be easily understood in
+        # the popo objects.
+        if isinstance(attr, Resource):
+            return attr.urn
+
+        # otherwise just emit as nothing to stop the looping.
+        return None
+
+    seen.append(attr)
+
+    # first check if the value is an actual dictionary.  If so, massage the values of it to deeply
+    # make sure this is a popo.
+    if isinstance(attr, dict):
+        result = {}
+        for key, value in attr.items():
+            # ignore private keys
+            if not key.startswith("_"):
+                result[key] = massage(value, seen)
+
+        return result
+
+    if isinstance(attr, Output):
+        return attr.apply(lambda v: massage(v, seen))
+
+    if hasattr(attr, "__dict__"):
+        # recurse on the dictionary itself.  It will be handled above.
+        return massage(attr.__dict__, seen)
+
+    # finally, recurse through iterables, converting into a list of massaged values.
+    return [massage(a, seen) for a in attr]
```

Comment:
Is it safe to do this at this point?  Are we sure that only iterable objects can get here?

Reply:
yes, it should be safe due to the work here aligning with the is_primitive check.  i.e. we only get primitives if we have dicts/dict-likes/iterables.  we handle the first two above.